File: datastructures-algorithms/src/simple/basic/questions/HeightOfBinaryTree.java
package simple.basic.questions;

public class HeightOfBinaryTree {

    static class Node{
        Node left,right;
        int data;
    }
    static Node newNode(int value){
        Node temp = new Node();
        temp.data = value;
        temp.left = temp.right = null;
        return temp;
    }

    public static void main(String[] args){
        System.out.println("Printing the height of Binary Tree");
        Node root = newNode(1);
        root.left = newNode(2);
        root.right = newNode(3);
        root.left.left = newNode(4);
        root.right.left = newNode(5);
        root.right.right = newNode(8);
        root.right.left.left = newNode(6);
        root.right.left.right = newNode(7);
        root.right.right.left = newNode(9);
        root.right.right.right = newNode(10);
        System.out.println("Height of Binary Tree ==>"+heightOfBinaryTree(null));
        System.out.println("Height of Binary Tree ==>"+heightOfBinaryTree(newNode(1)));
        System.out.println("Height of Binary Tree ==>"+heightOfBinaryTree(root));
    }

    public static int heightOfBinaryTree(Node root){
        if(root == null){
            return 0;
        }
        if(root.left == null && root.right ==null){
            return 1;
        }
        int left = heightOfBinaryTree(root.left);
        int right = heightOfBinaryTree(root.right);
        return (left>right)?left +1: right+1;
    }
}

File: datastructures-algorithms/src/dp/on/tree/HeightOfBinaryTree.scala
package dp.on.tree

object HeightOfBinaryTree {
  def main(args: Array[String]): Unit = {
    val tree: Node = Node.Node(10)

    val ltree: Node = Node.Node(2)
    val rtree: Node = Node.Node(3)

    val lltree: Node = Node.Node(23)
    val llltree: Node = Node.Node(35)

    tree.left = ltree
    tree.right = rtree

    tree.left.left = lltree
    tree.left.left.left = llltree

    println(tree)
    println("Height of Binary Tree ==> "+ heightOfBinaryTree(tree))

  }

  class Node(var data: Int)(implicit var left: Node, implicit var right: Node) {
    override def toString: String = "Node data => " + data + " Node left link " +
      "==>" +
      " " + left + "Node right link ==> " + right
  }

  implicit val lNode: Node = null
  implicit val rNode: Node = null

  object Node {
    def Node(nodeValue: Int) = new Node(nodeValue)(lNode, rNode)
  }

  def heightOfBinaryTree(rootNode: Node): Int = {

    if(rootNode == null) return 0
    val leftSubTree = heightOfBinaryTree((rootNode.left))
    val rightSubTree = heightOfBinaryTree(rootNode.right)

    return 1 + math.max(leftSubTree,rightSubTree)
  }

}
